quinn-proto/src/tests/mod.rs

```diff
@@ -1780,6 +1780,59 @@ fn congested_tail_loss() {
     pair.client_send(client_ch, s).write(&[42; 1024]).unwrap();
 }
 
+// Send a tail-loss probe when GSO segment_size is less than INITIAL_MTU
+#[test]
+fn tail_loss_small_segment_size() {
+    let _guard = subscribe();
+    let mut pair = Pair::default();
+    let (client_ch, server_ch) = pair.connect();
+
+    // No datagrams frames received in the handshake.
+    let server_stats = pair.server_conn_mut(server_ch).stats();
+    assert_eq!(server_stats.frame_rx.datagram, 0);
+
+    const DGRAM_LEN: usize = 1000; // Below INITIAL_MTU after packet overhead.
+    const DGRAM_NUM: usize = 5; // Enough to build a GSO batch.
```

Comment:
Does this need to be larger than 2?

Reply:
Yes, it needs to be 3 at least because the assertion happens in the next iteration when it is getting ready to build the 3rd datagram.  Rather than than making it fail exactly I thought adding a few datagrams would make it more reliable.  We want to make sure the entire batch works correctly.

Reply:
Actually, why wouldn't the issue be triggered after a *single* datagram? That should begin a GSO batch of size 1000-ish, and if a TLP is queued, that'll be the next packet, and end up too large.

Reply:
On the very first datagram it sets the size limit to 1200 because that is the maximum size it is allowed to produce.  Once the first datagram is populated however the first datagram is only 1030 bytes.  So when `next_datagram_size_limit` is being set for the 2nd datagram it is being set to 1030 because this is now the segment size.

The assert that notices the error happens earlier however, when congestion controller and pacing are checked if another datagram is still allows.  So the assertion failure only happens on the 3rd iteration.